File: C++/Patterns/Flyweight/Bitmap.cpp
#include "Bitmap.hpp"

namespace oobench {

    Bitmap::Bitmap(Screen& theScreen) : screen(theScreen) {
        pixels = new std::vector<PixelPoint>(screen.getSize());
    }

    Bitmap::~Bitmap() {
        delete pixels;
    }

    void Bitmap::show() {
        DrawContext context = DrawContext(screen);
        std::vector<PixelPoint>::iterator p = pixels->begin();

        while (p != pixels->end()) {
            (p++)->draw(context++);
        }
    }

    void Bitmap::showWithoutFlyweight() {
        char* scr = screen.getScreenPointer();
        std::vector<PixelPoint>::iterator p = pixels->begin();

        while (p != pixels->end()) {
            (*scr++) = (p++)->getPixel();
        }
    }
}

File: Java/org/oobench/patterns/flyweight/Bitmap.java
package org.oobench.patterns.flyweight;

import java.util.*;

public class Bitmap {
    private Screen screen;
    private PixelPoint[] pixels;

    public Bitmap(Screen theScreen) {
        screen = theScreen;
        int screenSize = screen.getSize();
        pixels = new PixelPoint[screenSize];
        for (int i = 0; i < screenSize; i++) {
            pixels[i] = new PixelPoint((byte)0);
        }
    }

    public void show() {
        DrawContext context = new DrawContext(screen);

        for (int i = 0; i < pixels.length; i++) {
            pixels[i].draw(context);
            context.next();
        }
    }

}
